api/envoy/api/v2/listener/listener.proto

```diff
@@ -88,6 +89,28 @@ message FilterChainMatch {
   // [#not-implemented-hide:]
   google.protobuf.UInt32Value suffix_len = 5;
 
+  enum RequestSourceType {
+    // Any request origin matches
+    ANY = 0;
+    // Match a request that originates from the same host network or namespace
+    // This would only match if the connecting and listener IPs are the same.
+    // See DISTINCT for an example.
+    SAME = 1;
+    // The request originates from an external network. This would usually be
+    // a non host local process.
+    // This condition will also match any local process communicating over
+    // a socket specifically bound to an interface different from the interface it
+    // contacts Envoy on.
+    // An example would be a host with an interface in 10.10.0.1/16 and Envoy
+    // listening on another interface 10.20.0.1/16. In this case if an application's
+    // client socket is explicitly bound to 10.10.0.1 the connection will
+    // effectively match the DISTINCT condition.
+    DISTINCT = 2;
+  }
+
+  // Specifies the source IP match type. Can be any, same or external network
```

Comment:
Can you elaborate on how this works with `source_prefix_ranges` (and `source_ports` if we decided to keep it, see https://github.com/envoyproxy/envoy/issues/4457.

Reply:
@PiotrSikora might want to weigh in, as he was originally defining the filter chain match semantics. I think you want to have source type as higher precedence than source CIDR range and that we can drop source port entirely. It also might make sense to give destination higher precedence than source CIDR range, since VIP and VIP ranges are the usual high order bit when reasoning about services.

Reply:
@nickolaev the changes in #4466 are wrong, since intermixing source and destination matching doesn't follow any rules. Source IPs and ports should be matched last (see #4457).

Due to the way the filter chain matching is implemented (i.e. matches on higher levels of the preference tree are binding even if there are no matches on the lower levels), I think that `source_type` should be the very last matching criteria, so that it doesn't take priority over source IP and/or port matching, although I imagine that people will use one or the other, so the ordering between source IP/port/type shouldn't matter too much.